collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -350,48 +366,192 @@ func (r *Receiver) anyStream(serverStream anyStreamServer, method string) (retEr
-	for {
-		// Receive a batch corresponding with one ptrace.Traces, pmetric.Metrics,
-		// or plog.Logs item.
-		req, err := serverStream.Recv()
+	doneCtx, doneCancel := context.WithCancel(streamCtx)
+	streamErrCh := make(chan error, 2)
+	pendingCh := make(chan batchResp, runtime.NumCPU())
+
+	go func() {
+		err := r.srvReceiveLoop(doneCtx, serverStream, pendingCh, method, ac)
+		streamErrCh <- err
+	}()
+
+	// WG is used to ensure main thread only returns once sender is finished sending responses for all requests.
+	var senderWG sync.WaitGroup
+	senderWG.Add(1)
+	go func() {
+		err := r.srvSendLoop(doneCtx, serverStream, pendingCh)
+		streamErrCh <- err
+		senderWG.Done()
+	}()
+
+	select {
+	case <-doneCtx.Done():
+		senderWG.Wait()
+		return status.Error(codes.Canceled, "server stream shutdown")
+	case retErr = <-streamErrCh:
+		doneCancel()
+		senderWG.Wait()
+		return
+	}
+}
+
+func (r *Receiver) recvOne(ctx context.Context, serverStream anyStreamServer, hrcv *headerReceiver, pendingCh chan<- batchResp, method string, ac arrowRecord.ConsumerAPI) (retErr error) {
+	r.inFlightWG.Add(1)
+	defer func() {
+		if retErr != nil {
+			r.inFlightWG.Done() // not processed
+			r.logStreamError(retErr)
+		}
+	}()
+
+	// Receive a batch corresponding with one ptrace.Traces, pmetric.Metrics,
+	// or plog.Logs item.
+	req, err := serverStream.Recv()
+
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return status.Error(codes.Canceled, "client stream shutdown")
+		} else if errors.Is(err, context.Canceled) {
+			return status.Error(codes.Canceled, "server stream shutdown")
+		}
+		return err
+	}
+
+	// Check for optional headers and set the incoming context.
+	thisCtx, authHdrs, err := hrcv.combineHeaders(ctx, req.GetHeaders())
+	if err != nil {
+		// Failing to parse the incoming headers breaks the stream.
+		return fmt.Errorf("arrow metadata error: %v", err)
+	}
+	var prevAcquiredBytes int
+	uncompSizeStr, sizeHeaderFound := authHdrs["otlp-pdata-size"]
+	if !sizeHeaderFound || len(uncompSizeStr) == 0 {
+		// This is a compressed size so make sure to acquire the difference when request is decompressed.
+		prevAcquiredBytes = proto.Size(req)
+	} else {
+		prevAcquiredBytes, err = strconv.Atoi(uncompSizeStr[0])
 		if err != nil {
-			// This includes the case where a client called CloseSend(), in
-			// which case we see an EOF error here.
-			r.logStreamError(err)
-
-			if errors.Is(err, io.EOF) {
-				return status.Error(codes.Canceled, "client stream shutdown")
-			} else if errors.Is(err, context.Canceled) {
-				return status.Error(codes.Canceled, "server stream shutdown")
+			return fmt.Errorf("failed to convert string to request size: %v", err)
+		}
+	}
+	// bounded queue to memory limit based on incoming uncompressed request size and waiters.
+	// Acquire will fail immediately if there are too many waiters,
+	// or will otherwise block until timeout or enough memory becomes available.
+	err = r.boundedQueue.Acquire(ctx, int64(prevAcquiredBytes))
```

Comment:
What prevents a buggy (or misbehaving) client/exporter from setting the 'otlp-pdata-size' arbitrarily high? Without an upper limit or some other form of protection, could this create a type of DoS attack at minimal cost? Could we detect when the 'otlp-pdata-size' does not match the actual size of the message after decompression and then ban the sender for a period of time? 

EDIT: Okay, there is detection in place with logging. Could we create some form of structured logs/events containing enough information to enable external (or internal) banning of the same clients that send us invalid data? The ban mechanism could be the topic of a future PR or be an external process triggered by these events.

Reply:
Thanks for the review! Yes we log in the case that the header `otlp-pdata-size`, but I think you are correct that we could face a DoS attack if we accept requests with headers that acquire a large portion of the semaphore. I can file an issue for this and would love to here more about how to get the appropriate client information that would allow us to block these clients.